runtime/compiler/x/codegen/J9CodeGenerator.cpp

```diff
@@ -403,6 +403,28 @@ J9::X86::CodeGenerator::suppressInliningOfRecognizedMethod(TR::RecognizedMethod
       }
    }
 
+bool
+J9::X86::CodeGenerator::supportsNonHelper(TR::SymbolReferenceTable::CommonNonhelperSymbol symbol)
+   {
+   bool result = false;
+
+   switch (symbol)
+      {
+      case TR::SymbolReferenceTable::atomicFetchAndAddSymbol:
+      case TR::SymbolReferenceTable::atomicSwapSymbol:
+         {
+         if (!self()->comp()->getOption(TR_DisableUnsafe) && !self()->comp()->compileRelocatableCode() && !TR::Compiler->om.canGenerateArraylets())
```

Comment:
IMO, these checks shouldn't belong to CodeGen. These conditions are to determine whether a Unsafe method should be transformed into NonHelper; they are NOT whether such NonHelper is supported. For example, other intrinsic methods may be transformed into the NonHelper but don't need checks for any of the tree condition.
In addition, the CodeGen actually do support for the two NonHelper, it is the ILGen or Optimizer that cannot transform the Unsafe calls to NonHelper under certain modes.

Reply:
I think I agree. I've greatly simplified things in 24ee533